File: ruby/run-length-encoding/run_length_encoding.rb
class RunLengthEncoding
  def self.encode(input)
    return '' if input.empty?

    input.chars.chunk { |letter| letter }.map do |k, v|
      "#{v.length if v.length > 1}#{k}"
    end.join
  end

  def self.decode(input)
    input.split(/(?<=[A-Za-z\s])/).map do |chunk|
      chunk = "1" + chunk if chunk.length == 1

      chars = chunk.chars
      chars.pop * chars.join.to_i
    end.join
  end
end

class BookKeeping
  VERSION = 3
end

File: typescript/run-length-encoding/run-length-encoding.ts
class RunLengthEncoding {
  static encode(input: string): string {
    const letters = input.split('');
    let encoded = '';
    let groupCount = 1;

    letters.forEach((letter, index) => {
      if (letter === letters[index + 1]) {
        groupCount++;
      } else {
        const character = `${groupCount > 1 ? groupCount : ''}${letter}`;
        encoded += character;
        groupCount = 1;
      }
    });

    return encoded;
  }

  static decode(input: string): string {
    const letters = input.split('');
    let decoded = '';
    let groupCount = '';

    letters.forEach(letter => {
      if (Number(letter)) {
        groupCount += letter;
      } else {
        const character =
          groupCount !== ''
            ? Array(Number(groupCount))
                .fill(letter)
                .join('')
            : letter;

        decoded += character;
        groupCount = '';
      }
    });

    return decoded;
  }
}

export default RunLengthEncoding;
